### ISSUE TITLE
Non-Prime Finite Fields

### ISSUE THREAD

**Issue by cathreya** [sympy/sympy#21104]
According to #9544 sympy still doesn't support finite fields of the form GF(p^m). I need to factorise and divide polynomials over GF(2^m). Would writing my own domain (or any other such hack) allow me to call existing functions for those?

**Comment by oscarbenjamin** [comment]
On master (not sure about 1.7.1) FiniteExtension is a domain that you could possibly use for this:
https://docs.sympy.org/dev/modules/polys/agca.html#finite-extensions

**Comment by jksuom** [comment]
Finite fields can be implemented as finite extensions of prime fields. The argument of `FiniteExtension` should be the minimal polynomial of a primitive element. That element will be the `generator` of the extension, and all non-zero elements will be its powers.
```
>>> from sympy.polys import Poly
>>> from sympy.polys.agca.extensions import FiniteExtension
>>> from sympy.abc import x
>>> K = FiniteExtension(Poly(x**3 + x + 1, x, modulus=2))
>>> a = K.generator
>>> for i in range(7):
...     print(a**i)
... 
1
x
x**2
x + 1
x**2 + x
x**2 + x + 1
x**2 + 1
>>> 
```
Most of the polynomial methods should be automatically available. It seems, however, that factorization is currently implemented (in `sympy.polys.galoistools`) only for polynomials over prime fields. For example, Berlecamp's Q-matrix, returned currently by `gf_Qmatrix`, has its entries in the prime field. The code should be improved.

**Comment by cathreya** [comment]
Thank you for your replies! Trying to declare a polynomial with domain as a FiniteExtension gives me `'MonogenicFiniteExtension' object has no attribute 'is_commutative'` and as jksuom mentioned, the galoistools implementation of factorization doesn't work. Is there any other way for me to be able to factorize?

**Comment by oscarbenjamin** [comment]
You need to try with master rather than 1.7.1

**Comment by cathreya** [comment]
@oscarbenjamin Right, my bad. On `1.8.dev` now. Now I am able to declare a polynomial, but my coefficients are only allowed to be 0 or 1. For example-
```
K = FiniteExtension(Poly(x**3 + x + 1, x, modulus=2))                                       
Poly(7+2*y,y,domain=K)  
```
Yields `Poly(1, y, domain='GF(2)[x]/(x**3 + x + 1)')`
(Here I assumed a coefficient i represents generator^i.)

I also tried the following:
```
a = K.generator
Poly(a**7+a**2*y,y,domain=K)
```
But this gives me an error: `TypeError: unsupported operand type(s) for *: 'ExtensionElement' and 'Symbol'`

**Comment by oscarbenjamin** [comment]
If you used `x` as the symbol for the extension then the coefficients should be in terms of `x`:
```python
In [11]: from sympy.polys.agca.extensions import FiniteExtension                                                                  

In [12]: K = FiniteExtension(Poly(x**3 + x + 1, x, modulus=2))                                                                    

In [13]: Poly((1 + x) + y*(1 + x), y, domain=K)                                                                                   
Out[13]: Poly((x + 1)*y + x + 1, y, domain='GF(2)[x]/(x**3 + x + 1)')

In [14]: a = Poly(x, y, domain=K)                                                                                                 

In [15]: for i in range(7): 
    ...:     print(a**i) 
    ...:                                                                                                                          
Poly(1, y, domain='GF(2)[x]/(x**3 + x + 1)')
Poly(x, y, domain='GF(2)[x]/(x**3 + x + 1)')
Poly(x**2, y, domain='GF(2)[x]/(x**3 + x + 1)')
Poly(x + 1, y, domain='GF(2)[x]/(x**3 + x + 1)')
Poly(x**2 + x, y, domain='GF(2)[x]/(x**3 + x + 1)')
Poly(x**2 + x + 1, y, domain='GF(2)[x]/(x**3 + x + 1)')
Poly(x**2 + 1, y, domain='GF(2)[x]/(x**3 + x + 1)')
```

**Comment by cathreya** [comment]
Thanks a lot for your reply! What is the state of multivariate functions over this currently? I tried the following:
```
In [40]: F = FiniteExtension(Poly(x**4 + x + 1, x, modulus=2))                                                                      
In [41]: Q = Poly(y**5*z**2 + (x**2 + x)*y**4*z**2 + x**3*y**2*z**2 + (x**3 + x**2 + x + 1)*y*z**2 + (x**3 + x**2 + x + 1)*z**2 + (x**3 + x**2)*z + x**2, y, z, domain='GF(2)[x]/(x**4 + x + 1)')
In [42]: p = Poly(x**2*y + 1, y, z, domain='GF(2)[x]/(x**4 + x + 1)')
In [43]: quo,rem = div(Q,p, domain=F)                                                                
```
This gives me a surprising error: `ExactQuotientFailed: x**2 does not divide 1 in GF(2)[x]`. Somehow the modulo irreducible is getting lost while performing the division?

`factor_list()` gives me a not implemented exception, and factor simply returns the polynomial as is.

**Comment by oscarbenjamin** [comment]
The FiniteExtension class was made into a domain in this PR https://github.com/sympy/sympy/pull/20631

There was some discussion about how to handle modulo division. I think that we were unsure in general what would work and what should be allowed so it was only made to work in limited situations.

**Comment by sylee957** [comment]
Can the finite field be constructed not only by using the polynomial of `GF(p)` coefficients, but also by using some `GF(p**n)` subfield coefficients?

**Comment by jksuom** [comment]
It should be possible to construct `GF(p**(n*m))` by means of an irreducible polynomial of degree `m` in `GF(p**n)[x]`.

**Comment by sylee957** [comment]
Yes, but the problem is how should we design a framework that allows the recursive construction like that.

**Comment by oscarbenjamin** [comment]
> Yes, but the problem is how should we design a framework that allows the recursive construction like that.

Is there any advantage in doing so?

**Comment by sylee957** [comment]
Sometimes I would need an option to work with relative extensions than absolute extensions.

For instance, sympy forces construction like `QQ(sqrt(2), sqrt(3))` as a absolute extension, but it can give a complicated presentation for `sqrt(2)`
```
>>> K = QQ.algebraic_field(sqrt(2)).algebraic_field(sqrt(3))
>>> print(K.to_sympy(K.convert(sqrt(2))))
-3*sqrt(3) + sqrt(2)**3/2 + 3*sqrt(3)*sqrt(2)**2/2
```

**Comment by oscarbenjamin** [comment]
That looks like a bug. It seems that `sqrt(2)` there is an `AlgebraicNumber` rather than a `Pow` which prevents it from simplifying automatically.

**Comment by sylee957** [comment]
No, my opinion is that naturally, the output format could be like ‘some polynomial of implicit radical’ than format that can be automatically simplified.
 
Even if it can be efficient to work with absolute extension, we need some algorithm to find an isomorphism to convert to other relative extensions.

**Comment by oscarbenjamin** [comment]
Yes, I have considered this. It's also slow to construct larger algebraic fields like:
```python
In [6]: %time K = QQ.algebraic_field(sqrt(2), sqrt(3), sqrt(5), sqrt(7), sqrt(11), sqrt(13))
^C---------------------------------------------------------------------------
KeyboardInterrupt
```
This is slow because of the time taken to compute the primitive element and calculations in such a domain would be slow because every element is a high degree polynomial even if it actually has a simple representation like `1 + sqrt(2)`.

I think that there could be a domain that represents algebraic number fields without using the primitive element theorem as e.g. `QQ<sqrt(2), sqrt(3), sqrt(5), ...>`. The internal representation would be something like:
```python
1 + sqrt(2)   --->   {1: 1, sqrt(2): 1}
```
Exact division would be possible using algorithms analogous to radsimp:
```python
In [7]: e = (1 + sqrt(2) + sqrt(3)) / (1 + sqrt(5) + sqrt(7))

In [8]: e
Out[8]: 
1 + √2 + √3
───────────
1 + √5 + √7

In [9]: radsimp(e)
Out[9]: 
-2⋅√105 - 2⋅√70 - 2⋅√35 - √21 - √14 - √7 + 3⋅√5 + 3⋅√10 + 11 + 3⋅√15 + 11⋅√2 + 11⋅√3
────────────────────────────────────────────────────────────────────────────────────
                                         19   
```
It would even be possible to make a domain like this in which it is possible to take square roots.

It is probably possible to generalise this approach to the idea of "sparse algebraic extensions" as discussed e.g. https://github.com/sympy/sympy/pull/20631#issuecomment-750386261

**Comment by patriciosanchezh** [comment]
What do we want the finite fields of non-prime order being able to do? just the common operations +-*/ and what else besides  having generated element? 

Do we want as a extension of the class FiniteField? 

Thanks in advance,
